File: java/src/main/be/steinsomers/bron_kerbosch/BronKerboschPivot.java
package be.steinsomers.bron_kerbosch;

import java.util.ArrayList;
import java.util.Collection;
import java.util.HashSet;
import java.util.Set;
import java.util.function.Consumer;
import java.util.stream.Collectors;

enum BronKerboschPivot {
    ;

    public static void explore(final UndirectedGraph graph, final Consumer<int[]> cliqueConsumer,
                               final PivotChoice pivotChoice) {
        final var order = graph.order();
        if (order > 0) {
            final var pivot = graph.maxDegreeVertex();
            // In this initial iteration, we don't need to represent the set of candidates
            // because all neighbours are candidates until excluded.
            final Set<Integer> mut_excluded = new HashSet<>(order);
            for (int v = 0; v < order; ++v) {
                final var neighbours = graph.neighbours(v);
                if (!neighbours.contains(pivot)) {
                    final var neighbouringExcluded = util.Intersect(neighbours, mut_excluded)
                            .collect(Collectors.toCollection(HashSet::new));
                    if (neighbouringExcluded.size() < neighbours.size()) {
                        final var neighbouringCandidates = util.Difference(neighbours, neighbouringExcluded)
                                .collect(Collectors.toCollection(HashSet::new));
                        visit(graph, cliqueConsumer, pivotChoice, neighbouringCandidates, neighbouringExcluded,
                                new int[]{v});
                    }
                    mut_excluded.add(v);
                }
            }
        }
    }

    public static void visit(
            final UndirectedGraph graph,
            final Consumer<int[]> cliqueConsumer,
            final PivotChoice pivotChoice,
            final Set<Integer> mut_candidates,
            final Set<Integer> mut_excluded,
            final int[] cliqueInProgress
    ) {
        assert mut_candidates.stream().allMatch(graph::hasDegree);
        assert mut_excluded.stream().allMatch(graph::hasDegree);
        assert util.AreDisjoint(mut_candidates, mut_excluded);
        //noinspection SizeReplaceableByIsEmpty
        assert mut_candidates.size() >= 1;
        if (mut_candidates.size() == 1) {
            // Same logic as below, stripped down for this common case
            final var v = mut_candidates.iterator().next();
            final var neighbours = graph.neighbours(v);
            if (util.AreDisjoint(neighbours, mut_excluded)) {
                cliqueConsumer.accept(util.Append(cliqueInProgress, v));
            }
        } else if (pivotChoice == PivotChoice.Arbitrary) {
            final var remainingCandidates = new ArrayList<>(mut_candidates);
            final int pivot = remainingCandidates.getFirst();
            visitAroundPivot(graph, cliqueConsumer, mut_candidates, mut_excluded, cliqueInProgress,
                    PivotChoice.Arbitrary, pivot, remainingCandidates);
        } else {
            visitMaxDegree(graph, cliqueConsumer, mut_candidates, mut_excluded, cliqueInProgress, pivotChoice);
        }
    }

    private static void visitMaxDegree(final UndirectedGraph graph, final Consumer<int[]> cliqueConsumer,
                                       final Set<Integer> mut_candidates, final Set<Integer> mut_excluded,
                                       final int[] cliqueInProgress, final PivotChoice pivotChoice) {
        assert pivotChoice == PivotChoice.MaxDegreeLocal || pivotChoice == PivotChoice.MaxDegreeLocalX;
        // Quickly handle locally unconnected candidates while finding pivot
        int pivot = -1;
        final Collection<Integer> remainingCandidates = new ArrayList<>(mut_candidates.size());
        long seenLocalDegree = 0;
        for (final var v : mut_candidates) {
            final var neighbours = graph.neighbours(v);
            final long localDegree = util.Intersect(neighbours, mut_candidates).count();
            if (localDegree == 0) {
                // Same logic as below, stripped down
                if (util.AreDisjoint(neighbours, mut_excluded)) {
                    cliqueConsumer.accept(util.Append(cliqueInProgress, v));
                }
            } else {
                if (seenLocalDegree < localDegree) {
                    seenLocalDegree = localDegree;
                    pivot = v;
                }
                remainingCandidates.add(v);
            }
        }
        if (pivotChoice == PivotChoice.MaxDegreeLocalX && !remainingCandidates.isEmpty()) {
            for (final var v : mut_excluded) {
                final var neighbours = graph.neighbours(v);
                final var localDegree = util.Intersect(neighbours, mut_candidates).count();
                if (seenLocalDegree < localDegree) {
                    seenLocalDegree = localDegree;
                    pivot = v;
                }
            }
        }
        visitAroundPivot(graph, cliqueConsumer, mut_candidates, mut_excluded, cliqueInProgress,
                pivotChoice, pivot, remainingCandidates);
    }

    private static void visitAroundPivot(final UndirectedGraph graph, final Consumer<int[]> cliqueConsumer,
                                         final Set<Integer> mut_candidates, final Set<Integer> mut_excluded,
                                         final int[] cliqueInProgress, final PivotChoice furtherPivotChoice,
                                         final int pivot, final Iterable<Integer> remainingCandidates) {
        for (final var v : remainingCandidates) {
            final var neighbours = graph.neighbours(v);
            if (!neighbours.contains(pivot)) {
                mut_candidates.remove(v);
                final var neighbouringCandidates = util.Intersect(neighbours, mut_candidates)
                        .collect(Collectors.toCollection(HashSet::new));
                if (!neighbouringCandidates.isEmpty()) {
                    final var neighbouringExcluded = util.Intersect(neighbours, mut_excluded)
                            .collect(Collectors.toCollection(HashSet::new));
                    visit(graph, cliqueConsumer, furtherPivotChoice,
                            neighbouringCandidates, neighbouringExcluded,
                            util.Append(cliqueInProgress, v));
                } else if (util.AreDisjoint(neighbours, mut_excluded)) {
                    cliqueConsumer.accept(util.Append(cliqueInProgress, v));
                }
                mut_excluded.add(v);
            }
        }
    }
}

File: kotlin/src/main/be/steinsomers/bron_kerbosch/BronKerboschPivot.kt
package be.steinsomers.bron_kerbosch

internal object BronKerboschPivot {
    fun explore(graph: UndirectedGraph, cliqueConsumer: CliqueConsumer, pivotChoice: PivotChoice) {
        val pivot = graph.maxDegreeVertices().firstOrNull()
        if (pivot != null) {
            // In this initial iteration, we don't need to represent the set of candidates
            // because all neighbours are candidates until excluded.
            val excluded = BooleanArray(graph.order)
            graph.connectedVertices().forEach { v ->
                val neighbours = graph.neighbours(v)
                if (!neighbours.contains(pivot)) {
                    val neighbouringExcluded = neighbours.filterTo(HashSet()) { v -> excluded[v.index] }
                    if (neighbouringExcluded.size < neighbours.size) {
                        val neighbouringCandidates = (neighbours subtract neighbouringExcluded).toMutableSet()
                        visit(
                            graph = graph, cliqueConsumer = cliqueConsumer,
                            pivotChoice = pivotChoice,
                            candidates = neighbouringCandidates, excluded = neighbouringExcluded,
                            clique = Clique.singleton(v)
                        )
                    }
                    excluded[v.index] = true
                }
            }
        }
    }

    fun visit(
        graph: UndirectedGraph,
        cliqueConsumer: CliqueConsumer,
        pivotChoice: PivotChoice,
        candidates: MutableSet<Vertex>,
        excluded: MutableSet<Vertex>,
        clique: Clique
    ) {
        Debug.assert { candidates.all(graph::hasDegree) }
        Debug.assert { excluded.all(graph::hasDegree) }
        Debug.assert { Util.areDisjoint(candidates, excluded) }
        Debug.assert { candidates.isNotEmpty() }
        if (candidates.size == 1) {
            // Same logic as below, stripped down for this common case
            val v = candidates.iterator().next()
            val neighbours = graph.neighbours(v)
            if (clique.size() + 1 >= cliqueConsumer.minSize && Util.areDisjoint(neighbours, excluded)) {
                cliqueConsumer.accept(clique.plus(v))
            }
        } else if (pivotChoice == PivotChoice.Arbitrary) {
            val remainingCandidates = ArrayList<Vertex>(candidates)
            val pivot = remainingCandidates[0]
            visitAroundPivot(
                graph, cliqueConsumer, candidates, excluded, clique, PivotChoice.Arbitrary,
                pivot, remainingCandidates
            )
        } else {
            visitMaxDegree(graph, cliqueConsumer, candidates, excluded, clique, pivotChoice)
        }
    }

    private fun visitMaxDegree(
        graph: UndirectedGraph, cliqueConsumer: CliqueConsumer,
        candidates: MutableSet<Vertex>, excluded: MutableSet<Vertex>,
        clique: Clique, pivotChoice: PivotChoice
    ) {
        require(pivotChoice == PivotChoice.MaxDegreeLocal || pivotChoice == PivotChoice.MaxDegreeLocalX)
        // Quickly handle locally unconnected candidates while finding pivot
        var pivot = Vertex(0)
        val remainingCandidates: MutableCollection<Vertex> = ArrayList(candidates.size)
        var seenLocalDegree = 0
        for (v in candidates) {
            val neighbours = graph.neighbours(v)
            val localDegree = Util.overlap(neighbours, candidates)
            if (localDegree == 0) {
                // Same logic as below, stripped down
                if (clique.size() + 1 >= cliqueConsumer.minSize && Util.areDisjoint(neighbours, excluded)) {
                    cliqueConsumer.accept(clique.plus(v))
                }
            } else {
                if (seenLocalDegree < localDegree) {
                    seenLocalDegree = localDegree
                    pivot = v
                }
                remainingCandidates.add(v)
            }
        }
        if (pivotChoice == PivotChoice.MaxDegreeLocalX && remainingCandidates.isNotEmpty()) {
            for (v in excluded) {
                val neighbours = graph.neighbours(v)
                val localDegree = Util.overlap(neighbours, candidates)
                if (seenLocalDegree < localDegree) {
                    seenLocalDegree = localDegree
                    pivot = v
                }
            }
        }
        visitAroundPivot(
            graph, cliqueConsumer, candidates, excluded, clique, pivotChoice, pivot, remainingCandidates
        )
    }

    private fun visitAroundPivot(
        graph: UndirectedGraph, cliqueConsumer: CliqueConsumer,
        candidates: MutableSet<Vertex>, excluded: MutableSet<Vertex>,
        clique: Clique, furtherPivotChoice: PivotChoice,
        pivot: Vertex, remainingCandidates: Iterable<Vertex>
    ) {
        for (v in remainingCandidates) {
            val neighbours = graph.neighbours(v)
            if (!neighbours.contains(pivot)) {
                candidates.remove(v)
                val neighbouringCandidates = Util.intersect(neighbours, candidates)
                if (neighbouringCandidates.isNotEmpty()) {
                    val neighbouringExcluded = Util.intersect(neighbours, excluded)
                    visit(
                        graph = graph, cliqueConsumer = cliqueConsumer,
                        pivotChoice = furtherPivotChoice,
                        candidates = neighbouringCandidates, excluded = neighbouringExcluded,
                        clique = clique.plus(v)
                    )
                } else if (clique.size() + 1 >= cliqueConsumer.minSize && Util.areDisjoint(
                        neighbours,
                        excluded
                    )
                ) {
                    cliqueConsumer.accept(clique.plus(v))
                }
                excluded.add(v)
            }
        }
    }
}
